File: safeguard-backend/app/services/socket_service.py
import socketio
from typing import Dict, Any
import asyncio
import json

class SocketManager:
    def __init__(self):
        self.sio = socketio.AsyncServer(
            cors_allowed_origins=["http://localhost:3000"],
            async_mode='asgi'
        )
        self.connected_users: Dict[str, str] = {}  # user_id: session_id
        self.user_sessions: Dict[str, str] = {}    # session_id: user_id
        
        self.setup_handlers()
    
    def setup_handlers(self):
        @self.sio.event
        async def connect(sid, environ):
            print(f"Client connected: {sid}")
        
        @self.sio.event
        async def disconnect(sid):
            print(f"Client disconnected: {sid}")
            if sid in self.user_sessions:
                user_id = self.user_sessions[sid]
                del self.connected_users[user_id]
                del self.user_sessions[sid]
        
        @self.sio.event
        async def authenticate(sid, data):
            """Authenticate WebSocket connection"""
            try:
                from app.core.security import verify_token
                token = data.get('token')
                user_id = verify_token(token)
                
                if user_id:
                    self.connected_users[user_id] = sid
                    self.user_sessions[sid] = user_id
                    await self.sio.emit('authenticated', {'success': True}, sid)
                else:
                    await self.sio.emit('authentication_error', {'error': 'Invalid token'}, sid)
            except Exception as e:
                await self.sio.emit('authentication_error', {'error': str(e)}, sid)
        
        @self.sio.event
        async def location_update(sid, data):
            """Handle location updates"""
            if sid not in self.user_sessions:
                return
            
            user_id = self.user_sessions[sid]
            latitude = data.get('latitude')
            longitude = data.get('longitude')
            accuracy = data.get('accuracy')
            
            # Update location in database (you'll need to inject db session)
            # For now, just broadcast to nearby users
            nearby_update = {
                'user_id': user_id,
                'location': {
                    'latitude': latitude,
                    'longitude': longitude,
                    'accuracy': accuracy
                },
                'timestamp': data.get('timestamp')
            }
            
            # Broadcast to all connected users (in production, only to nearby users)
            await self.sio.emit('location_broadcast', nearby_update)
        
        @self.sio.event
        async def voice_phrase_detected(sid, data):
            """Handle voice phrase detection"""
            if sid not in self.user_sessions:
                return
            
            user_id = self.user_sessions[sid]
            transcript = data.get('transcript', '')
            confidence = data.get('confidence', 0.0)
            
            # Emit emergency trigger event
            await self.sio.emit('emergency_trigger', {
                'type': 'voice',
                'user_id': user_id,
                'transcript': transcript,
                'confidence': confidence,
                'timestamp': data.get('timestamp')
            }, sid)
    
    async def emit_to_user(self, user_id: str, event: str, data: Any):
        """Emit event to specific user"""
        if user_id in self.connected_users:
            sid = self.connected_users[user_id]
            await self.sio.emit(event, data, sid)
    
    async def broadcast(self, event: str, data: Any):
        """Broadcast event to all connected users"""
        await self.sio.emit(event, data)
    
    def get_asgi_app(self):
        """Get ASGI app for FastAPI integration"""
        return socketio.ASGIApp(self.sio)

File: safeguard-frontend/src/services/socketService.js
import io from 'socket.io-client';
import { SOCKET_URL } from '../config/api';

class SocketService {
  constructor() {
    this.socket = null;
    this.isConnected = false;
    this.listeners = new Map();
  }

  connect() {
    if (this.socket) {
      this.disconnect();
    }

    this.socket = io(SOCKET_URL, {
      transports: ['websocket', 'polling'],
      timeout: 20000,
    });

    this.socket.on('connect', () => {
      console.log('✅ Connected to server');
      this.isConnected = true;
      this.authenticate();
    });

    this.socket.on('disconnect', (reason) => {
      console.log('❌ Disconnected from server:', reason);
      this.isConnected = false;
    });

    this.socket.on('connect_error', (error) => {
      console.error('❌ Connection error:', error);
      this.isConnected = false;
    });

    this.socket.on('authenticated', (data) => {
      console.log('✅ Socket authenticated:', data);
    });

    this.socket.on('authentication_error', (error) => {
      console.error('❌ Authentication error:', error);
    });

    // Emergency events
    this.socket.on('emergency_alert', (data) => {
      console.log('🚨 Emergency alert received:', data);
      this.emit('emergency_alert_received', data);
    });

    this.socket.on('emergency_dismissed', (data) => {
      console.log('✅ Emergency dismissed:', data);
      this.emit('emergency_dismissed_received', data);
    });

    this.socket.on('emergency_trigger', (data) => {
      console.log('🚨 Emergency trigger:', data);
      this.emit('emergency_trigger_received', data);
    });

    // Location events
    this.socket.on('location_broadcast', (data) => {
      console.log('📍 Location update received:', data);
      this.emit('location_update_received', data);
    });

    this.socket.on('nearby_users_update', (data) => {
      console.log('👥 Nearby users update:', data);
      this.emit('nearby_users_updated', data);
    });

    return this.socket;
  }

  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
      this.isConnected = false;
    }
  }

  authenticate() {
    const token = localStorage.getItem('token');
    if (token && this.socket) {
      this.socket.emit('authenticate', { token });
    }
  }

  // Event emitters for server
  updateLocation(latitude, longitude, accuracy = null) {
    if (this.socket && this.isConnected) {
      this.socket.emit('location_update', {
        latitude,
        longitude,
        accuracy,
        timestamp: new Date().toISOString(),
      });
    }
  }

  detectVoicePhrase(transcript, confidence) {
    if (this.socket && this.isConnected) {
      this.socket.emit('voice_phrase_detected', {
        transcript,
        confidence,
        timestamp: new Date().toISOString(),
      });
    }
  }

  // Local event management
  on(event, callback) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    this.listeners.get(event).push(callback);
  }

  off(event, callback) {
    if (this.listeners.has(event)) {
      const callbacks = this.listeners.get(event);
      const index = callbacks.indexOf(callback);
      if (index > -1) {
        callbacks.splice(index, 1);
      }
    }
  }

  emit(event, data) {
    if (this.listeners.has(event)) {
      this.listeners.get(event).forEach(callback => callback(data));
    }
  }
}

export default new SocketService();